source/_posts/rails-api-backend-for-backbone.markdown

```diff
@@ -0,0 +1,322 @@
+
+```
+Prefix Verb   URI Pattern          Controller#Action
+ todos GET    /todos(.:format)     todos#index
+       POST   /todos(.:format)     todos#create
+  todo GET    /todos/:id(.:format) todos#show
+       PATCH  /todos/:id(.:format) todos#update
+       PUT    /todos/:id(.:format) todos#update
+       DELETE /todos/:id(.:format) todos#destroy
+```
+
+Do not forget to run `bin/rake db:migrate`, so the database is ready when the time comes to test our application.
+
+### Defining how TODO items should be serialized
+
+Our only-API application will respond incoming requests in JSON format, and here is where Active Model Serializer plays an important role. It requires to define the TodoSerializer class and provide the list of attributes from Todo model to include in the responses.
+
+Active Model Serializer offers a command to generate this serializer:
+
+```
+bin/rails g serializer todo title completed order
+```
+
+The result is the file `app/serializers/todo_serializer.rb`:
+
+```ruby
+class TodoSerializer < ActiveModel::Serializer
+  attributes :id, :title, :completed, :order
+end
+```
+
+At this point, we have implemented with almost zero effort a working backend application. We can run `bin/rails s` to start the web server and test our API with the help of `curl` command.
+
+Let's create a new Todo
+
+```
+curl -H "Content-Type:application/json; charset=utf-8" -d '{"title":"something to do","order":1,"completed":false}' http://localhost:3000/todos
+```
+
+and it should return:
+
+```
+{"id":1,"title":"something to do","completed":false,"order":1}
+```
+
+Now, we can try to get the list of TODOs:
+
+```
+curl http://localhost:3000/todos
+```
+
+and the result should be:
+
+```
+[{"id":1,"title":"something to do","completed":false,"order":1}]
+```
+
+### Putting both components to work together
+
+It's time to integrate the Backbone application with our backend implementation!
+
+The client side TODO list application that we are using uses the browser local storage, so we need to indicate our API application is the new storage for the data.
+
+Let's replace the following lines in `js/collections/todos.js`:
+
+```js
+  // Save all of the todo items under the `"todos"` namespace.
+  localStorage: new Backbone.LocalStorage('todos-backbone'),
+```
+
+with a definition for the URL of our API endpoint for TODO items:
+
+```js
+  // Set the rails-api backend endpoint for this specific model
+  url: 'http://localhost:3000/todos',
+```
+
+After this change, we are almost ready. Our last outstanding task is to configure the cross origin policy to make possible the communication between both components. This is necessary when the backend and the client components are in different domains. Since we are doing local testing, we will run both things in localhost but using different ports, so we will end up having CORS errors in the browser if we don't configure any cross origin policy.
```

Comment:
make possible the communication between both components -> make communication between both components possible.